service/history/replication/sync_state_retriever.go

```diff
@@ -376,9 +389,37 @@ func (s *SyncStateRetrieverImpl) getSnapshot(mutableState workflow.MutableState)
 	return mutableStateProto, nil
 }
 
-func (s *SyncStateRetrieverImpl) getEventsBlob(ctx context.Context, branchToken []byte, startEventId int64, endEventId int64) ([]*commonpb.DataBlob, error) {
+func (s *SyncStateRetrieverImpl) getEventsBlob(
+	ctx context.Context,
+	workflowKey definition.WorkflowKey,
+	versionHistory *history.VersionHistory,
+	startEventId int64,
+	endEventId int64,
+) ([]*commonpb.DataBlob, error) {
+	var eventBlobs []*commonpb.DataBlob
+
+	if s.eventBlobCache != nil {
+		for {
+			eventVersion, err := versionhistory.GetVersionHistoryEventVersion(versionHistory, startEventId)
+			if err != nil {
+				return nil, err
+			}
+			xdcCacheValue, ok := s.eventBlobCache.Get(persistence.NewXDCCacheKey(workflowKey, startEventId, eventVersion))
+			if !ok {
+				break
+			}
+			left := endEventId - startEventId
+			if int64(len(xdcCacheValue.EventBlobs)) >= left {
+				eventBlobs = append(eventBlobs, xdcCacheValue.EventBlobs[:left]...)
```

Comment:
Don't do this truncation, we always need to ship the full batch. 
I think that's how the newRun case right now on L312, we say the endEventID is 2 but what we actually want is the full batch.

Reply:
We can add some sanity check to make sure this only happens for newRun (i.e. startEventID  =  1), then we don't need to worry about syncing extra events.